src/a11y-dialog.js

```diff
@@ -1,283 +1,237 @@
- * @return {A11yDialog}
- */
-A11yDialog.prototype.show = function (event) {
-  // If the dialog is already open, abort
-  if (this.shown) {
-    return this
-  }
+/** @type A11yDialogType */
+export default class A11yDialog {
+  /** @param {element} HTMLElement */
+  constructor(element) {
+    this.$el = element
+    this.id = this.$el.getAttribute('data-a11y-dialog') || this.$el.id
 
-  // Keep a reference to the currently focused element to be able to restore
-  // it later
-  this._previouslyFocused = document.activeElement
-  this.$el.removeAttribute('aria-hidden')
-  this.shown = true
+    this.$el.setAttribute('aria-hidden', 'true')
+    this.$el.setAttribute('aria-modal', 'true')
+    this.$el.setAttribute('tabindex', '-1')
 
-  // Set the focus to the dialog element
-  moveFocusToDialog(this.$el)
+    if (!this.$el.hasAttribute('role')) {
+      this.$el.setAttribute('role', 'dialog')
+    }
 
-  // Bind a focus event listener to the body element to make sure the focus
-  // stays trapped inside the dialog while open, and start listening for some
-  // specific key presses (TAB and ESC)
-  document.body.addEventListener('focus', this._maintainFocus, true)
-  document.addEventListener('keydown', this._bindKeypress)
+    // Keep a collection of dialog openers, each of which will be bound a click
+    // event listener to open the dialog
+    this.openers = $$(`[data-a11y-dialog-show="${this.id}"]`)
+    this.openers.forEach(opener => {
+      opener.addEventListener('click', this.show)
+    })
+
+    // Keep a collection of dialog closers, each of which will be bound a click
+    // event listener to close the dialog
+    this.closers = $$(
+      `[data-a11y-dialog="${this.id}"] [data-a11y-dialog-hide], #${this.id} [data-a11y-dialog-hide], [data-a11y-dialog-hide="${this.id}"]`
+    )
+    this.closers.forEach(closer => {
+      closer.addEventListener('click', this.hide)
+    })
 
-  // Execute all callbacks registered for the `show` event
-  this._fire('show', event)
+    this.listeners = {}
 
-  return this
-}
+    this.previouslyFocused = null
+    this.shown = false
 
-/**
- * Hide the dialog element, restore the focus to the previously active element,
- * stop listening for some specific key presses and fire all registered
- * callbacks for `hide` event
- *
- * @param {CustomEvent} event - An optional `CustomEvent`, passed to `hide`
- *                              event listeners
- * @return {A11yDialog}
- */
-A11yDialog.prototype.hide = function (event) {
-  // If the dialog is already closed, abort
-  if (!this.shown) {
-    return this
+    // Execute all callbacks registered for the `create` event
+    this.fire('create')
   }
 
-  this.shown = false
-  this.$el.setAttribute('aria-hidden', 'true')
+  /**
+   * Destroy the current instance (after making sure the dialog has been hidden)
+   * and remove all associated listeners from dialog openers and closers
+   */
+  destroy = () => {
```

Comment:
Can we maybe do `destroy () {` instead? I can’t remember the difference between the two. I think one does implicit `this` binding?

Reply:
I cannot remember if instance fields (`destroy = () => {}`) implicitly bind this or simply do not create a new lexical scope in the first place, but either way this behavior is what we want. If we switch to non-arrow functions, tests start failing because we need to bind again.

Reply:
Lovely, thanks. :)